draft-ietf-mls-protocol.md

```diff
@@ -1435,44 +1465,230 @@ all arrive at the following state:
+The primary purpose of the handshake protocol is to provide an authenticated
+group key exchange to participants. In order to protect Application messages
+sent among those participants, the Application secret provided by the Handshake
+key schedule is used to derive encryption keys for the Message Protection Layer.
+
+Application messages MUST be protected with the Authenticated-Encryption
+with Associated-Data (AEAD) encryption scheme associated with the MLS ciphersuite.
+Note that "Authenticated" in this context does not mean messages are known to
+be sent by a specific participant but only from a legitimate member of the group.
+To authenticate a message from a particular member, signatures are required.
+Handshake messages MUST use asymmetric signatures
+to strongly authenticate the sender of a message; Application messages SHOULD
+use the signature scheme to provide the same property.
+
+Each participant maintains their own chain of Application secrets, where the first
+one is derived based on a secret chained from the Epoch secret.
+As shown in {{key-schedule}}, the initial Application secret is bound to the
+identity of each participant to avoid collisions and allow support for decryption
+of reordered messages.
+
+Subsequent Application secrets MUST be rotated for each Application message, independently
+from the Group secret updates, in order to provide stronger cryptographic
+security properties for messages. To later encrypt and decrypt Application messages,
+we use this secret to derive AEAD encryption key and nonce.
+In all cases, a participant MUST NOT encrypt more than expected by the AEAD scheme
+with Nonces and keys generated from the same Application secrets.
+
+[[OPEN ISSUE: The AEAD key exhaustion bound applies on a per Application
+Secret basis. This depends on the number of messages per Application secrets
+that we allow. Should we set a limit ?
+]]
+
+Note that each change to the Group through a Handshake message will cause
+a change of the Group Secret. Hence this change MUST be applied before encrypting
+any new Application message. This is required for confidentiality reasons
+in order for Members to avoid receiving messages from the group after leaving or
+being added to, or excluded from the Group.
+
+## Application Key Schedule {#key-schedule-application}
+
+Updating the Application secret and deriving the associated AEAD key and nonce can
+be summarized as the following Application key schedule where
+each participant's Application secret chain looks as follows after the initial
+derivation:
 
-The primary purpose of this protocol is to enable an authenticated
-group key exchange among participants.  In order to protect messages sent among
-those participants, an application will need to specify how messages are protected.
+~~~~~
+           application_secret_N-1
+                     |
+                     +--> HKDF-Expand-Label(.,"nonce", "", nonce_length)
+                     |    = write_nonce_N-1_[sender]
```

Comment:
If you use the nonce directly like this rather than as an IV mask on message seq, then you absolutely can never use the same keys twice.

Reply:
Well, this way you would get key reuse.